waku/waku_api/rest/builder.nim

```diff
@@ -0,0 +1,185 @@
+import
+  ../../waku/common/utils/nat,
+  ../../waku/waku_node,
+  ../../waku/discovery/waku_discv5,
+  ../../waku/factory/external_config,
+  ../../waku/waku_api/message_cache,
+  ../../waku/waku_api/handlers,
+  ../../waku/waku_api/rest/server,
+  ../../waku/waku_api/rest/debug/handlers as rest_debug_api,
+  ../../waku/waku_api/rest/relay/handlers as rest_relay_api,
+  ../../waku/waku_api/rest/filter/handlers as rest_filter_api,
+  ../../waku/waku_api/rest/lightpush/handlers as rest_lightpush_api,
+  ../../waku/waku_api/rest/store/handlers as rest_store_api,
+  ../../waku/waku_api/rest/health/handlers as rest_health_api,
+  ../../waku/waku_api/rest/admin/handlers as rest_admin_api,
+  ../../waku/waku_core/topics
+
+## Monitoring and external interfaces
+
+# Used to register api endpoints that are not currently installed as keys,
+# values are holding error messages to be returned to the client
+# NOTE: {.threadvar.} is used to make the global variable GC safe for the closure uses it
+# It will always be called from main thread anyway.
+# Ref: https://nim-lang.org/docs/manual.html#threads-gc-safety
+var restServerNotInstalledTab {.threadvar.}: TableRef[string, string]
+restServerNotInstalledTab = newTable[string, string]()
+
+proc startRestServerEsentials*(
+    nodeHealthMonitor: WakuNodeHealthMonitor, conf: WakuNodeConf
+): Result[WakuRestServerRef, string] =
+  let requestErrorHandler: RestRequestErrorHandler = proc(
+      error: RestRequestError, request: HttpRequestRef
+  ): Future[HttpResponseRef] {.async: (raises: [CancelledError]).} =
+    try:
+      case error
+      of RestRequestError.Invalid:
+        return await request.respond(Http400, "Invalid request", HttpTable.init())
+      of RestRequestError.NotFound:
+        let paths = request.rawPath.split("/")
+        let rootPath =
+          if len(paths) > 1:
+            paths[1]
+          else:
+            ""
+        restServerNotInstalledTab[].withValue(rootPath, errMsg):
+          return await request.respond(Http404, errMsg[], HttpTable.init())
+        do:
+          return await request.respond(
+            Http400,
+            "Bad request initiated. Invalid path or method used.",
+            HttpTable.init(),
+          )
+      of RestRequestError.InvalidContentBody:
+        return await request.respond(Http400, "Invalid content body", HttpTable.init())
+      of RestRequestError.InvalidContentType:
+        return await request.respond(Http400, "Invalid content type", HttpTable.init())
+      of RestRequestError.Unexpected:
+        return defaultResponse()
+    except HttpWriteError:
+      error "Failed to write response to client", error = getCurrentExceptionMsg()
+      discard
+
+    return defaultResponse()
+
+  let allowedOrigin =
+    if len(conf.restAllowOrigin) > 0:
+      some(conf.restAllowOrigin.join(","))
+    else:
+      none(string)
+
+  let address = conf.restAddress
+  let port = Port(conf.restPort + conf.portsShift)
+  let server =
```

Comment:
I think in case conf.rest is false we shall not create the restServer.

Reply:
> I think in case conf.rest is false we shall not create the restServer.

Good point and thanks for the comment! Nevertheless, I'm trying to respect the current implementation where we always create, at least, one ` restServer` when calling `startRestServerEsentials`:

https://github.com/waku-org/nwaku/blob/6d135b0d1b853cbcc59aa18b0d6de961e3b6e948/apps/wakunode2/wakunode2.nim#L130

https://github.com/waku-org/nwaku/blob/6d135b0d1b853cbcc59aa18b0d6de961e3b6e948/waku/factory/app.nim#L333